### ISSUE TITLE
Flaky tests

### ISSUE THREAD

**Issue by EnricoMi** [horovod/horovod#2414]
Issue to collect flaky unit tests.

The Unit Test Results comment reports failed tests:
![grafik](https://user-images.githubusercontent.com/44700269/97816236-2bc4fe00-1c94-11eb-914e-31459a1c8e5d.png)

... but Buildkite succeeds:
![grafik](https://user-images.githubusercontent.com/44700269/97816269-6d55a900-1c94-11eb-9688-a1f7b52fc676.png)

This is because failed test job reran and eventually succeeds, hence a flaky test.

- https://buildkite.com/horovod/horovod/builds/4239#4bc39604-926e-4120-9937-35220da79547
  https://github.com/horovod/horovod/runs/1339365236
1 out of 6 runs failed: test_hosts_added_and_removed (test.integration.test_elastic_torch.ElasticTorchTests)
test-results/factory-Elastic-Tests-test-cpu-gloo-py3_6-tf1_15_0-keras2_2_4-torch1_2_0/artifacts/junit.gloo.elastic.xml
```
self = <test_elastic_torch.ElasticTorchTests testMethod=test_hosts_added_and_removed>
mock_get_min_start_hosts = <MagicMock name='_get_min_start_hosts' id='140507299633528'>

    @mock.patch('horovod.runner.elastic.driver.DISCOVER_HOSTS_FREQUENCY_SECS', 0.01)
    @mock.patch('horovod.runner.gloo_run._get_min_start_hosts', return_value=1)
    def test_hosts_added_and_removed(self, mock_get_min_start_hosts):
        for slots, np, min_np, max_np in [(2, 2, 2, 4), (1, 1, 1, 2)]:
            discovery_schedule = [
                (0, ['localhost:{}'.format(slots)]),
                (1, ['localhost:{}'.format(slots), '127.0.0.1:{}'.format(slots)]),
                (None, ['127.0.0.1:{}'.format(slots)]),
            ]
    
            results = self._run(discovery_schedule, np=np, min_np=min_np, max_np=max_np)
    
            assert len(results) == 3
    
            assert results[0]['start_rank'] == 0
            assert results[0]['size'] == slots
            assert results[0]['hostname'] == 'localhost'
    
            assert results[1]['start_rank'] == 0
            assert results[1]['size'] == slots * 2
            assert results[1]['hostname'] == 'localhost'
    
>           assert results[2]['start_rank'] == slots
E           AssertionError

elastic_common.py:138: AssertionError
```

**Comment by tgaddair** [comment]
Hmm, it would be useful to see which rank is getting assigned rank 0 here instead of `slots`.  If it's `slots+1`, then that would indicate some kind of non-determinism in the assignment of slots to ranks by the driver.  If it's `0`, then that would imply that the first worker wasn't properly removed (or the timing wasn't synced correctly in the test).

Do you know if the param values are available in the log somewhere?  If not, I can put together a PR to add additional diagnostic info.

**Comment by tgaddair** [comment]
This should be fixed by #2524, but we should keep an eye out as the flaky nature of the failures means it may resurface in the future.

Will close for now.
